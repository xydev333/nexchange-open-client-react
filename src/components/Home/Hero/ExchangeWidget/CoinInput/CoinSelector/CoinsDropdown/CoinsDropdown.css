.coin-currency-dropdown {
  min-width: 250px;
  border-radius: 3px;
  box-shadow: 0 16px 38px -12px rgba(0, 0, 0, 0.56), 0 4px 25px 0px rgba(0, 0, 0, 0.12), 0 8px 10px -5px rgba(0, 0, 0, 0.2);
  -webkit-transition: all 100ms linear;
  -moz-transition: all 100ms linear;
  -o-transition: all 100ms linear;
  -ms-transition: all 100ms linear;
  transition: all 100ms linear;
  position: absolute;
  right: 0;
  top: 61px;
  z-index: 500;
  background: #fff;
}

.coins-list {
  max-height: 233px;
  padding: 0 15px;
  overflow-y: scroll;
  overflow-x: hidden;
}

.coins-search {
  position: relative;
  border-bottom: 1px solid #dedede;
  padding: 0 15px;
}

input {
  width: 200px;
  border: none;
  padding: 10px 0 10px 15px;
  font-size: 16px;
}

.clear.clear {
  font-size: 18px;
  position: absolute;
  top: 11px;
  right: 10px;
  cursor: pointer;
  display: none;
}

.clear.active {
  display: block;
}

.coin {
  width: 250px;
  margin-top: 10px;
  margin-bottom: 10px;
  cursor: pointer;
}

.coin i {
  font-size: 20px;
}
