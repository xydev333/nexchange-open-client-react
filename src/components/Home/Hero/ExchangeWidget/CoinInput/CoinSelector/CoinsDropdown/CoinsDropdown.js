import React, { Component } from 'react';
import _ from 'lodash';
import Fuse from 'fuse.js';
import cx from 'classnames';
import urlParams from 'Utils/urlParams';
import debounce from 'Utils/debounce';
import styles from './CoinsDropdown.css';
import { I18n } from 'react-i18next';

class CoinsDropdown extends Component {
  state = { value: '' };

  componentDidMount = () => {
    this.searchInput.focus();
  };

  handleChange = event => {
    this.setState({ value: event.target.value });
  };

  handleSubmit = event => {
    event.preventDefault();
    event.stopPropagation();

    const coins = this.getCoins();

    if (coins[0]) {
      this.props.onClick(coins[0].code);
    }
  };

  clear = () => {
    this.setState({ value: '' });
  };

  trackEvent = debounce(coinSearched => {
    window.ga('send', {
      hitType: 'event',
      eventCategory: 'Coins dropdown search',
      eventAction: this.props.type,
      eventValue: coinSearched,
    });
  }, 300);

  searchCoins = coins => {
    if (!this.state.value) return coins;

    const fuse = new Fuse(coins, {
      shouldSort: true,
      threshold: 0.2,
      keys: ['code', 'name'],
    });

    this.trackEvent(this.state.value);

    return fuse.search(this.state.value);
  };

  getCoins = () => {
    const params = urlParams();
    let filteredCoins = this.props.coinsInfo.filter(coin => {
      if (params && params.hasOwnProperty('test')) {
        return this.props.type.toUpperCase() === 'DEPOSIT' ? coin.is_quote_of_enabled_pair_for_test : coin.is_base_of_enabled_pair_for_test;
      }

      return this.props.type.toUpperCase() === 'DEPOSIT' ? coin.is_quote_of_enabled_pair : coin.is_base_of_enabled_pair;
    });
    filteredCoins = _.sortBy(filteredCoins, 'is_crypto');
    filteredCoins = this.searchCoins(filteredCoins);

    return filteredCoins;
  };

  renderSearch = () => {
    return (
    <I18n ns="translations">
     {(t) => (
      <form className={styles['coins-search']} onSubmit={this.handleSubmit}>
        <i className="fas fa-search" aria-hidden="true" />
        <input
          type="text"
          placeholder={t('generalterms.search')}
          ref={input => (this.searchInput = input)}
          onChange={this.handleChange}
          value={this.state.value}
        />
        <i className={`material-icons ${this.state.value ? cx(styles.clear, styles.active) : styles.clear}`} onClick={this.clear}>
          clear
        </i>
      </form>
      )}
	 </I18n>
    );
  };

  renderCoins = () => {
    return (
      <div className={styles['coins-list']}>
        {this.getCoins().map(coin => (
          <div className={`row ${styles.coin}`} key={coin.code} onClick={() => this.props.onClick(coin.code)}>
            <div className="col-xs-3 text-center">
              <i className={`cc ${coin.code} ${coin.code}`} />
            </div>
            <div className="col-xs-3">
              <b>{coin.code}</b>
            </div>
            <div className="col-xs-6 text-capitalize">{coin.name}</div>
          </div>
        ))}
      </div>
    );
  };

  render = () => {
    return (
      <div className={`coin-currency-dropdown ${styles['coin-currency-dropdown']}`}>
        {this.renderSearch()}
        {this.renderCoins()}
      </div>
    );
  };
}

export default CoinsDropdown;
