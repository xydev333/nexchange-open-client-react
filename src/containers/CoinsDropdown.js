import React, { Component } from 'react';
import _ from 'lodash';
import urlParams from '../helpers/urlParams';
import Fuse from 'fuse.js';
import '../css/_coins-dropdown.scss';

class CoinsDropdown extends Component {
  constructor(props) {
    super(props);
    this.state = { value: '' };
  }

  componentDidMount = () => {
    this.searchInput.focus();
  };

  handleChange = event => {
    this.setState({ value: event.target.value });
  };

  searchCoins = filteredCoins => {
    if (!this.state.value) return filteredCoins;

    const fuse = new Fuse(filteredCoins, {
      shouldSort: true,
      threshold: 0.2,
      keys: ['code', 'name'],
    });

    return fuse.search(this.state.value);
  };

  clear = () => {
    this.setState({ value: '' });
  };

  renderSearch() {
    return (
      <div className="coins-search">
        <i class="fa fa-search" aria-hidden="true" />
        <input
          type="text"
          placeholder="Search"
          ref={input => (this.searchInput = input)}
          onChange={this.handleChange}
          value={this.state.value}
        />
        <i className={`material-icons clear ${this.state.value ? 'active' : null}`} onClick={this.clear}>
          clear
        </i>
      </div>
    );
  }

  renderCoins() {
    const params = urlParams();
    let filteredCoins = this.props.coinsInfo.filter(coin => {
      if (params && params.hasOwnProperty('test')) {
        return this.props.type.toUpperCase() === 'DEPOSIT' ? coin.is_quote_of_enabled_pair_for_test : coin.is_base_of_enabled_pair_for_test;
      }

      return this.props.type.toUpperCase() === 'DEPOSIT' ? coin.is_quote_of_enabled_pair : coin.is_base_of_enabled_pair;
    });
    filteredCoins = _.sortBy(filteredCoins, 'is_crypto');
    filteredCoins = this.searchCoins(filteredCoins);

    return (
      <div className="coins-list">
        {filteredCoins.map(coin => (
          <div className="row coin" key={coin.code} onClick={() => this.props.onClick(coin.code)}>
            <div className="col-xs-3">
              <b>{coin.code}</b>
            </div>
            <div className="col-xs-3 text-center">
              <i className={`cc ${coin.code} ${coin.code}`} />
            </div>
            <div className="col-xs-6 text-capitalize">{coin.name}</div>
          </div>
        ))}
      </div>
    );
  }

  render() {
    return (
      <div className="coin-currency-dropdown">
        {this.renderSearch()}
        {this.renderCoins()}
      </div>
    );
  }
}

export default CoinsDropdown;
