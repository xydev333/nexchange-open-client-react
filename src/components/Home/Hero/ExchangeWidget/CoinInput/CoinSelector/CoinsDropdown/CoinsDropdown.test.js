import React from 'react';
import { shallow } from 'enzyme';
import CoinsDropdown from './CoinsDropdown.js';
import coinsInfo from 'Mocks/currency.js';

describe('CoinsDropdown', () => {
  let wrapShallowDeposit, wrapShallowReceive;

  beforeEach(() => {
    wrapShallowDeposit = shallow(<CoinsDropdown type="deposit" onClick={jest.fn()} coinsInfo={coinsInfo} />);
    wrapShallowReceive = shallow(<CoinsDropdown type="receive" onClick={jest.fn()} coinsInfo={coinsInfo} />);
  });

  it('renders correctly', () => {
    expect(wrapShallowDeposit).toMatchSnapshot();
    expect(wrapShallowReceive).toMatchSnapshot();
  });

  it('deposit dropdown contains correct coins', () => {
    for (const coin of coinsInfo) {
      if (coin.is_quote_of_enabled_pair) {
        expect(wrapShallowDeposit.find(`[data-test="${coin.code}"]`).length).toEqual(1);
      } else {
        expect(wrapShallowDeposit.find(`[data-test="${coin.code}"]`).length).toEqual(0);
      }
    }
  });

  it('receive dropdown contains correct coins', () => {
    for (const coin of coinsInfo) {
      if (coin.is_base_of_enabled_pair) {
        expect(wrapShallowReceive.find(`[data-test="${coin.code}"]`).length).toEqual(1);
      } else {
        expect(wrapShallowReceive.find(`[data-test="${coin.code}"]`).length).toEqual(0);
      }
    }
  });
});
