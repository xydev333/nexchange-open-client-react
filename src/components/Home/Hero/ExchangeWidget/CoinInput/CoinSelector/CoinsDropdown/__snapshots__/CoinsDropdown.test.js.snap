// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinsDropdown renders correctly 1`] = `ShallowWrapper {}`;

exports[`CoinsDropdown renders correctly 2`] = `ShallowWrapper {}`;
